File: examples/node/hal_query.ts
/**
 * HAL Query Example
 *
 * Query HAL (Hardware Abstraction Layer) pins, signals, and parameters.
 * Useful for debugging HAL configurations and monitoring I/O.
 *
 * Usage:
 *   npx tsx hal_query.ts pins "axis.*"
 *   npx tsx hal_query.ts signals
 *   npx tsx hal_query.ts components
 *   npx tsx hal_query.ts watch "spindle.0.speed-out" "axis.x.pos-cmd"
 */

import { program } from "commander";
import {
  HalServiceClient,
  GetSystemStatusRequest,
  Metadata,
  QueryPinsCommand,
  QuerySignalsCommand,
  QueryParamsCommand,
  QueryComponentsCommand,
  WatchRequest,
  HalValue,
  HalType,
  PinDirection,
  ParamDirection,
  halTypeToJSON,
  pinDirectionToJSON,
  credentials,
  status,
} from "linuxcnc-grpc";
import type { ServiceError } from "linuxcnc-grpc";

program
  .option("--host <host>", "gRPC server host", "localhost")
  .option("--port <port>", "gRPC server port", "50051")
  .option("--interval <interval>", "Update interval in ms for watch command", "500");

program
  .command("pins [pattern]")
  .description("Query HAL pins")
  .action((pattern) => runCommand("pins", pattern || "*"));

program
  .command("signals [pattern]")
  .description("Query HAL signals")
  .action((pattern) => runCommand("signals", pattern || "*"));

program
  .command("params [pattern]")
  .description("Query HAL parameters")
  .action((pattern) => runCommand("params", pattern || "*"));

program
  .command("components [pattern]")
  .description("Query HAL components")
  .action((pattern) => runCommand("components", pattern || "*"));

program
  .command("watch <names...>")
  .description("Watch values for changes")
  .action((names) => runCommand("watch", names));

program
  .command("status")
  .description("Get HAL system status")
  .action(() => runCommand("status", ""));

const RPC_DEADLINE = 5000;

// Declare client before parse() since actions may run synchronously during parse()
let client: HalServiceClient;

program.parse();

function getAddress(): string {
  const opts = program.opts();
  return `${opts.host}:${opts.port}`;
}

function getIntervalMs(): number {
  const opts = program.opts();
  return parseInt(opts.interval);
}

function formatValue(value: HalValue | undefined): string {
  if (!value) return "?";
  if (value.bitValue !== undefined) {
    return value.bitValue ? "TRUE" : "FALSE";
  }
  if (value.floatValue !== undefined) {
    return value.floatValue.toPrecision(6);
  }
  if (value.s32Value !== undefined) {
    return value.s32Value.toString();
  }
  if (value.u32Value !== undefined) {
    return value.u32Value.toString();
  }
  if (value.s64Value !== undefined) {
    return value.s64Value.toString();
  }
  if (value.u64Value !== undefined) {
    return value.u64Value.toString();
  }
  if (value.portValue !== undefined) {
    return value.portValue;
  }
  return "?";
}

function formatType(halType: HalType): string {
  return halTypeToJSON(halType).replace("HAL_", "");
}

function formatDirection(direction: PinDirection): string {
  return pinDirectionToJSON(direction)
    .replace("HAL_", "")
    .replace("PIN_DIR_", "");
}

function queryPins(pattern: string): void {
  const request = QueryPinsCommand.create();
  request.pattern = pattern;

  client.queryPins(request, new Metadata(), { deadline: new Date(Date.now() + RPC_DEADLINE) }, (err, response) => {
    if (err) {
      console.error(`Error: ${err.message}`);
      client.close();
      process.exit(1);
    }
    if (!response.success) {
      console.error(`Error: ${response.error}`);
      process.exit(1);
    }

    console.log(`Found ${response.pins.length} pins matching '${pattern}':\n`);
    console.log(
      `${"Name".padEnd(50)} ${"Type".padEnd(6)} ${"Dir".padEnd(4)} ${"Value".padEnd(15)} Signal`
    );
    console.log("-".repeat(90));

    const pins = [...response.pins].sort((a, b) => a.name.localeCompare(b.name));
    for (const pin of pins) {
      const direction = formatDirection(pin.direction);
      const value = formatValue(pin.value);
      const pinType = formatType(pin.type);
      const signal = pin.signal || "-";
      console.log(
        `${pin.name.padEnd(50)} ${pinType.padEnd(6)} ${direction.padEnd(4)} ${value.padEnd(15)} ${signal}`
      );
    }
    client.close();
  });
}

function querySignals(pattern: string): void {
  const request = QuerySignalsCommand.create();
  request.pattern = pattern;

  client.querySignals(request, new Metadata(), { deadline: new Date(Date.now() + RPC_DEADLINE) }, (err, response) => {
    if (err) {
      console.error(`Error: ${err.message}`);
      client.close();
      process.exit(1);
    }
    if (!response.success) {
      console.error(`Error: ${response.error}`);
      process.exit(1);
    }

    console.log(`Found ${response.signals.length} signals matching '${pattern}':\n`);
    console.log(
      `${"Name".padEnd(40)} ${"Type".padEnd(6)} ${"Value".padEnd(15)} ${"Driver".padEnd(30)} Readers`
    );
    console.log("-".repeat(100));

    const signals = [...response.signals].sort((a, b) =>
      a.name.localeCompare(b.name)
    );
    for (const sig of signals) {
      const value = formatValue(sig.value);
      const sigType = formatType(sig.type);
      const driver = sig.driver || "(none)";
      const readers = sig.readerCount > 0 ? `${sig.readerCount} readers` : "-";
      console.log(
        `${sig.name.padEnd(40)} ${sigType.padEnd(6)} ${value.padEnd(15)} ${driver.padEnd(30)} ${readers}`
      );
    }
    client.close();
  });
}

function queryParams(pattern: string): void {
  const request = QueryParamsCommand.create();
  request.pattern = pattern;

  client.queryParams(request, new Metadata(), { deadline: new Date(Date.now() + RPC_DEADLINE) }, (err, response) => {
    if (err) {
      console.error(`Error: ${err.message}`);
      client.close();
      process.exit(1);
    }
    if (!response.success) {
      console.error(`Error: ${response.error}`);
      process.exit(1);
    }

    console.log(`Found ${response.params.length} parameters matching '${pattern}':\n`);
    console.log(
      `${"Name".padEnd(50)} ${"Type".padEnd(6)} ${"Mode".padEnd(4)} Value`
    );
    console.log("-".repeat(80));

    const params = [...response.params].sort((a, b) =>
      a.name.localeCompare(b.name)
    );
    for (const param of params) {
      const value = formatValue(param.value);
      const paramType = formatType(param.type);
      const mode = param.direction === ParamDirection.HAL_RW ? "RW" : "RO";
      console.log(
        `${param.name.padEnd(50)} ${paramType.padEnd(6)} ${mode.padEnd(4)} ${value}`
      );
    }
    client.close();
  });
}

function queryComponents(pattern: string): void {
  const request = QueryComponentsCommand.create();
  request.pattern = pattern;

  client.queryComponents(request, new Metadata(), { deadline: new Date(Date.now() + RPC_DEADLINE) }, (err, response) => {
    if (err) {
      console.error(`Error: ${err.message}`);
      client.close();
      process.exit(1);
    }
    if (!response.success) {
      console.error(`Error: ${response.error}`);
      process.exit(1);
    }

    console.log(`Found ${response.components.length} components matching '${pattern}':\n`);
    console.log(
      `${"Name".padEnd(30)} ${"ID".padEnd(6)} ${"Ready".padEnd(6)} ${"Pins".padEnd(6)} Params`
    );
    console.log("-".repeat(60));

    const comps = [...response.components].sort((a, b) =>
      a.name.localeCompare(b.name)
    );
    for (const comp of comps) {
      const ready = comp.ready ? "Yes" : "No";
      console.log(
        `${comp.name.padEnd(30)} ${comp.id.toString().padEnd(6)} ${ready.padEnd(6)} ${comp.pins.length.toString().padEnd(6)} ${comp.params.length}`
      );
    }
    client.close();
  });
}

function watchValues(names: string[]): void {
  const request = WatchRequest.create();
  request.names = names;
  request.intervalMs = getIntervalMs();

  console.log(`Watching ${names.length} values (interval: ${getIntervalMs()}ms)`);
  console.log("Press Ctrl+C to stop\n");

  const stream = client.watchValues(request);

  stream.on("data", (batch) => {
    for (const change of batch.changes) {
      const oldVal = formatValue(change.oldValue);
      const newVal = formatValue(change.newValue);
      const ts = new Date(Number(change.timestamp) / 1000000).toLocaleTimeString();
      console.log(`[${ts}] ${change.name}: ${oldVal} -> ${newVal}`);
    }
  });

  stream.on("error", (err: ServiceError) => {
    if (err.code === status.CANCELLED) {
      // Normal cancellation
    } else {
      console.error(`\ngRPC error: ${err.code}: ${err.details}`);
      client.close();
      process.exit(1);
    }
  });

  stream.on("end", () => {
    client.close();
  });

  process.on("SIGINT", () => {
    stream.cancel();
  });
}

function getSystemStatus(): void {
  client.getSystemStatus(GetSystemStatusRequest.create(), new Metadata(), { deadline: new Date(Date.now() + RPC_DEADLINE) }, (err, status) => {
    if (err) {
      console.error(`Error: ${err.message}`);
      client.close();
      process.exit(1);
    }

    console.log("HAL System Status");
    console.log("=".repeat(40));
    console.log(`Pins:       ${status.pins.length}`);
    console.log(`Signals:    ${status.signals.length}`);
    console.log(`Parameters: ${status.params.length}`);
    console.log(`Components: ${status.components.length}`);
    console.log(`Simulation: ${status.isSim}`);
    console.log(`Real-time:  ${status.isRt}`);
    console.log(`Userspace:  ${status.isUserspace}`);
    if (status.kernelVersion) {
      console.log(`Kernel:     ${status.kernelVersion}`);
    }
    client.close();
  });
}

function runCommand(command: string, arg: string | string[]): void {
  const address = getAddress();
  console.error(`Connecting to ${address}...`);
  client = new HalServiceClient(address, credentials.createInsecure(), {
    'grpc.initial_reconnect_backoff_ms': 1000,
    'grpc.max_reconnect_backoff_ms': 5000,
  });

  switch (command) {
    case "pins":
      queryPins(arg as string);
      break;
    case "signals":
      querySignals(arg as string);
      break;
    case "params":
      queryParams(arg as string);
      break;
    case "components":
      queryComponents(arg as string);
      break;
    case "watch":
      watchValues(arg as string[]);
      break;
    case "status":
      getSystemStatus();
      break;
    default:
      console.error(`Unknown command: ${command}`);
      process.exit(1);
  }
}

File: examples/rust/src/bin/hal_query.rs
//! HAL Query Example
//!
//! Query HAL (Hardware Abstraction Layer) pins, signals, and parameters.
//! Useful for debugging HAL configurations and monitoring I/O.
//!
//! # Usage
//!
//! ```bash
//! cargo run --bin hal_query -- pins "axis.*"
//! cargo run --bin hal_query -- signals
//! cargo run --bin hal_query -- components
//! cargo run --bin hal_query -- watch spindle.0.speed-out axis.x.pos-cmd
//! ```

use clap::{Parser, Subcommand};
use futures_util::StreamExt;
use linuxcnc_grpc::hal::hal_service_client::HalServiceClient;
use linuxcnc_grpc::hal::{
    GetSystemStatusRequest, HalType, HalValue, ParamDirection, PinDirection, QueryComponentsCommand,
    QueryParamsCommand, QueryPinsCommand, QuerySignalsCommand, WatchRequest,
};
use std::time::{Duration, UNIX_EPOCH};
use tokio::signal;
use tokio::time::timeout;

#[derive(Parser, Debug)]
#[command(name = "hal_query")]
#[command(about = "Query HAL pins, signals, parameters, and components")]
struct Args {
    /// gRPC server host
    #[arg(long, default_value = "localhost")]
    host: String,

    /// gRPC server port
    #[arg(long, default_value_t = 50051)]
    port: u16,

    /// Update interval in ms for watch command
    #[arg(long, default_value_t = 500)]
    interval: i32,

    #[command(subcommand)]
    command: Command,
}

#[derive(Subcommand, Debug)]
enum Command {
    /// Query HAL pins
    Pins {
        /// Glob pattern to match pin names
        #[arg(default_value = "*")]
        pattern: String,
    },
    /// Query HAL signals
    Signals {
        /// Glob pattern to match signal names
        #[arg(default_value = "*")]
        pattern: String,
    },
    /// Query HAL parameters
    Params {
        /// Glob pattern to match parameter names
        #[arg(default_value = "*")]
        pattern: String,
    },
    /// Query HAL components
    Components {
        /// Glob pattern to match component names
        #[arg(default_value = "*")]
        pattern: String,
    },
    /// Watch values for changes
    Watch {
        /// Names of pins/signals/params to watch
        names: Vec<String>,
    },
    /// Get HAL system status
    Status,
}

fn format_value(value: &Option<HalValue>) -> String {
    match value {
        Some(v) => match &v.value {
            Some(linuxcnc_grpc::hal::hal_value::Value::BitValue(b)) => {
                if *b { "TRUE" } else { "FALSE" }.to_string()
            }
            Some(linuxcnc_grpc::hal::hal_value::Value::FloatValue(f)) => format!("{:.6}", f),
            Some(linuxcnc_grpc::hal::hal_value::Value::S32Value(i)) => format!("{}", i),
            Some(linuxcnc_grpc::hal::hal_value::Value::U32Value(u)) => format!("{}", u),
            Some(linuxcnc_grpc::hal::hal_value::Value::S64Value(i)) => format!("{}", i),
            Some(linuxcnc_grpc::hal::hal_value::Value::U64Value(u)) => format!("{}", u),
            Some(linuxcnc_grpc::hal::hal_value::Value::PortValue(s)) => s.clone(),
            None => "?".to_string(),
        },
        None => "?".to_string(),
    }
}

fn format_type(hal_type: i32) -> &'static str {
    match HalType::try_from(hal_type) {
        Ok(HalType::HalBit) => "BIT",
        Ok(HalType::HalFloat) => "FLOAT",
        Ok(HalType::HalS32) => "S32",
        Ok(HalType::HalU32) => "U32",
        Ok(HalType::HalS64) => "S64",
        Ok(HalType::HalU64) => "U64",
        Ok(HalType::HalPort) => "PORT",
        _ => "?",
    }
}

fn format_direction(direction: i32) -> &'static str {
    match PinDirection::try_from(direction) {
        Ok(PinDirection::HalIn) => "IN",
        Ok(PinDirection::HalOut) => "OUT",
        Ok(PinDirection::HalIo) => "IO",
        _ => "?",
    }
}

async fn query_pins(
    client: &mut HalServiceClient<tonic::transport::Channel>,
    pattern: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let response = client
        .query_pins(tonic::Request::new(QueryPinsCommand {
            pattern: pattern.to_string(),
        }))
        .await?
        .into_inner();

    if !response.success {
        eprintln!("Error: {}", response.error);
        return Ok(());
    }

    println!("Found {} pins matching '{}':\n", response.pins.len(), pattern);
    println!(
        "{:<50} {:<6} {:<4} {:<15} {}",
        "Name", "Type", "Dir", "Value", "Signal"
    );
    println!("{}", "-".repeat(90));

    let mut pins = response.pins;
    pins.sort_by(|a, b| a.name.cmp(&b.name));

    for pin in pins {
        let direction = format_direction(pin.direction);
        let value = format_value(&pin.value);
        let pin_type = format_type(pin.r#type);
        let signal = if pin.signal.is_empty() {
            "-"
        } else {
            &pin.signal
        };
        println!(
            "{:<50} {:<6} {:<4} {:<15} {}",
            pin.name, pin_type, direction, value, signal
        );
    }

    Ok(())
}

async fn query_signals(
    client: &mut HalServiceClient<tonic::transport::Channel>,
    pattern: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let response = client
        .query_signals(tonic::Request::new(QuerySignalsCommand {
            pattern: pattern.to_string(),
        }))
        .await?
        .into_inner();

    if !response.success {
        eprintln!("Error: {}", response.error);
        return Ok(());
    }

    println!(
        "Found {} signals matching '{}':\n",
        response.signals.len(),
        pattern
    );
    println!(
        "{:<40} {:<6} {:<15} {:<30} {}",
        "Name", "Type", "Value", "Driver", "Readers"
    );
    println!("{}", "-".repeat(100));

    let mut signals = response.signals;
    signals.sort_by(|a, b| a.name.cmp(&b.name));

    for sig in signals {
        let value = format_value(&sig.value);
        let sig_type = format_type(sig.r#type);
        let driver = if sig.driver.is_empty() {
            "(none)"
        } else {
            &sig.driver
        };
        let readers = if sig.reader_count > 0 {
            format!("{} readers", sig.reader_count)
        } else {
            "-".to_string()
        };
        println!(
            "{:<40} {:<6} {:<15} {:<30} {}",
            sig.name, sig_type, value, driver, readers
        );
    }

    Ok(())
}

async fn query_params(
    client: &mut HalServiceClient<tonic::transport::Channel>,
    pattern: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let response = client
        .query_params(tonic::Request::new(QueryParamsCommand {
            pattern: pattern.to_string(),
        }))
        .await?
        .into_inner();

    if !response.success {
        eprintln!("Error: {}", response.error);
        return Ok(());
    }

    println!(
        "Found {} parameters matching '{}':\n",
        response.params.len(),
        pattern
    );
    println!("{:<50} {:<6} {:<4} {}", "Name", "Type", "Mode", "Value");
    println!("{}", "-".repeat(80));

    let mut params = response.params;
    params.sort_by(|a, b| a.name.cmp(&b.name));

    for param in params {
        let value = format_value(&param.value);
        let param_type = format_type(param.r#type);
        let mode = match ParamDirection::try_from(param.direction) {
            Ok(ParamDirection::HalRw) => "RW",
            _ => "RO",
        };
        println!("{:<50} {:<6} {:<4} {}", param.name, param_type, mode, value);
    }

    Ok(())
}

async fn query_components(
    client: &mut HalServiceClient<tonic::transport::Channel>,
    pattern: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let response = client
        .query_components(tonic::Request::new(QueryComponentsCommand {
            pattern: pattern.to_string(),
        }))
        .await?
        .into_inner();

    if !response.success {
        eprintln!("Error: {}", response.error);
        return Ok(());
    }

    println!(
        "Found {} components matching '{}':\n",
        response.components.len(),
        pattern
    );
    println!(
        "{:<30} {:<6} {:<6} {:<6} {}",
        "Name", "ID", "Ready", "Pins", "Params"
    );
    println!("{}", "-".repeat(60));

    let mut comps = response.components;
    comps.sort_by(|a, b| a.name.cmp(&b.name));

    for comp in comps {
        let ready = if comp.ready { "Yes" } else { "No" };
        println!(
            "{:<30} {:<6} {:<6} {:<6} {}",
            comp.name,
            comp.id,
            ready,
            comp.pins.len(),
            comp.params.len()
        );
    }

    Ok(())
}

async fn watch_values(
    client: &mut HalServiceClient<tonic::transport::Channel>,
    names: Vec<String>,
    interval_ms: i32,
) -> Result<(), Box<dyn std::error::Error>> {
    let request = tonic::Request::new(WatchRequest {
        names: names.clone(),
        interval_ms,
    });

    let mut stream = client.watch_values(request).await?.into_inner();

    println!("Watching {} values (interval: {}ms)", names.len(), interval_ms);
    println!("Press Ctrl+C to stop\n");

    let ctrl_c = async {
        signal::ctrl_c()
            .await
            .expect("Failed to install Ctrl+C handler");
    };

    tokio::pin!(ctrl_c);

    loop {
        tokio::select! {
            _ = &mut ctrl_c => {
                break;
            }
            result = timeout(Duration::from_secs(30), stream.next()) => {
                match result {
                    Ok(Some(Ok(batch))) => {
                        for change in batch.changes {
                            let old_val = format_value(&change.old_value);
                            let new_val = format_value(&change.new_value);
                            let ts = UNIX_EPOCH + Duration::from_nanos(change.timestamp as u64);
                            let datetime = chrono_lite_format(ts);
                            println!("[{}] {}: {} -> {}", datetime, change.name, old_val, new_val);
                        }
                    }
                    Ok(Some(Err(e))) => {
                        eprintln!("Stream error: {}", e);
                        break;
                    }
                    Ok(None) => {
                        break;
                    }
                    Err(_) => {
                        eprintln!("Timeout: no update received in 30 seconds");
                        break;
                    }
                }
            }
        }
    }

    Ok(())
}

fn chrono_lite_format(ts: std::time::SystemTime) -> String {
    let duration = ts.duration_since(UNIX_EPOCH).unwrap_or_default();
    let total_secs = duration.as_secs();

    // Days since Unix epoch to Y/M/D (simplified Gregorian)
    let mut days = (total_secs / 86400) as i64;
    let secs_of_day = total_secs % 86400;
    let hours = secs_of_day / 3600;
    let minutes = (secs_of_day % 3600) / 60;
    let seconds = secs_of_day % 60;

    // Compute year/month/day from days since 1970-01-01 (UTC)
    let mut year: i64 = 1970;
    loop {
        let days_in_year = if year % 4 == 0 && (year % 100 != 0 || year % 400 == 0) { 366 } else { 365 };
        if days < days_in_year {
            break;
        }
        days -= days_in_year;
        year += 1;
    }
    let leap = year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
    let days_in_months = [31, if leap { 29 } else { 28 }, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
    let mut month = 0usize;
    for &dim in &days_in_months {
        if days < dim {
            break;
        }
        days -= dim;
        month += 1;
    }

    format!("{:04}-{:02}-{:02} {:02}:{:02}:{:02} UTC", year, month + 1, days + 1, hours, minutes, seconds)
}

async fn get_system_status(
    client: &mut HalServiceClient<tonic::transport::Channel>,
) -> Result<(), Box<dyn std::error::Error>> {
    let response = client
        .get_system_status(tonic::Request::new(GetSystemStatusRequest {}))
        .await?
        .into_inner();

    println!("HAL System Status");
    println!("{}", "=".repeat(40));
    println!("Pins:       {}", response.pins.len());
    println!("Signals:    {}", response.signals.len());
    println!("Parameters: {}", response.params.len());
    println!("Components: {}", response.components.len());
    println!("Simulation: {}", response.is_sim);
    println!("Real-time:  {}", response.is_rt);
    println!("Userspace:  {}", response.is_userspace);
    if !response.kernel_version.is_empty() {
        println!("Kernel:     {}", response.kernel_version);
    }

    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();
    let addr = format!("http://{}:{}", args.host, args.port);

    let mut client = HalServiceClient::connect(addr).await?;

    match args.command {
        Command::Pins { pattern } => query_pins(&mut client, &pattern).await?,
        Command::Signals { pattern } => query_signals(&mut client, &pattern).await?,
        Command::Params { pattern } => query_params(&mut client, &pattern).await?,
        Command::Components { pattern } => query_components(&mut client, &pattern).await?,
        Command::Watch { names } => {
            if names.is_empty() {
                eprintln!("Error: watch command requires at least one name");
                std::process::exit(1);
            }
            watch_values(&mut client, names, args.interval).await?
        }
        Command::Status => get_system_status(&mut client).await?,
    }

    Ok(())
}
